src/flow/assembly_lmh_old.hh

```diff
@@ -28,598 +47,688 @@
+        this->used_fields_ += eq_fields_->cross_section;
+        this->used_fields_ += eq_fields_->conductivity;
+        this->used_fields_ += eq_fields_->anisotropy;
+        this->used_fields_ += eq_fields_->sigma;
+        this->used_fields_ += eq_fields_->water_source_density;
+        this->used_fields_ += eq_fields_->extra_source;
+        this->used_fields_ += eq_fields_->storativity;
+        this->used_fields_ += eq_fields_->extra_storativity;
+        this->used_fields_ += eq_fields_->bc_type;
+        this->used_fields_ += eq_fields_->bc_pressure;
+        this->used_fields_ += eq_fields_->bc_flux;
+        this->used_fields_ += eq_fields_->bc_robin_sigma;
+        this->used_fields_ += eq_fields_->bc_switch_pressure;
     }
 
-    /** Loads the local system from a map: element index -> LocalSystem,
-     * if it exits, or if the full solution is not yet reconstructed,
-     * and reconstructs the full solution on the element.
-     * Currently used only for seepage BC.
-     */
-    void load_local_system(const DHCellAccessor& dh_cell)
-    {
-        // do this only once per element
-        if(bc_fluxes_reconstruted)
-            return;
+    /// Destructor.
+    virtual ~MHMatrixAssemblyLMH() {}
 
-        // possibly find the corresponding local system
-        auto ls = ad_->seepage_bc_systems.find(dh_cell.elm_idx());
-        if (ls != ad_->seepage_bc_systems.end())
-        {
-            arma::vec schur_solution = ad_->p_edge_solution.get_subvec(loc_schur_.row_dofs);            
-            // reconstruct the velocity and pressure
-            ls->second.reconstruct_solution_schur(schur_offset_, schur_solution, reconstructed_solution_);
+    /// Initialize auxiliary vectors and other data members
+    void initialize(ElementCacheMap *element_cache_map) {
+        //this->balance_ = eq_data_->balance_;
+        this->element_cache_map_ = element_cache_map;
 
-            postprocess_velocity(dh_cell, reconstructed_solution_);
+        fe_ = std::make_shared< FE_P_disc<dim> >(0);
+        fe_values_side_.initialize(*this->quad_low_, *fe_, update_normal_vectors);
 
-            bc_fluxes_reconstruted = true;
-        }
+        fe_values_.initialize(quad_rt_, fe_rt_, update_values | update_JxW_values | update_quadrature_points);
+
+        ndofs_ = fe_values_.n_dofs();
+        qsize_ = fe_values_.n_points();
+        this->use_dirichlet_switch_ = true;
+
+        // local numbering of dofs for MH system
+        // note: this shortcut supposes that the fe_system is the same on all elements
+        // TODO the function should be DiscreteSpace.fe(ElementAccessor)
+        // and correct form fe(ad_->dh_->own_range().begin()->elm()) (see documentation of DiscreteSpace::fe)
+        auto fe = eq_data_->dh_->ds()->fe()[Dim<dim>{}];
+        FESystem<dim>* fe_system = dynamic_cast<FESystem<dim>*>(fe.get());
+        eq_data_->loc_side_dofs[dim-1] = fe_system->fe_dofs(0);
```

Comment:
These could rather be local variables of the Assembly or even just local variables.

Reply:
These 'dofs' variables are defined in EqData as std::array of size 3 and hold one item for each dimension. Values are 'global' on all elements of one dimension. Data are used in begin() method too where we iterate through elements of different dimensions and therefore we need to preset data of all dimensions.